dask/dataframe/tests/test_pyarrow.py

```diff
@@ -0,0 +1,164 @@
+import numpy as np
+import pandas as pd
+import pytest
+
```

Comment:
It may be good to add a gpu-ci test for the case that we are mixing the "cudf" backend with `"dataframe.convert_string"` set to `True`. I don't *think* this should be a problem, but seems important to test, given that dask-cudf uses the same `_Frame.__init__`

Note that I'm about to test this myself, but need to set up a fresh RAPIDS environment.

Reply:
Ah - Seems that something like this will currently fail:

```python
    with dask.config.set(
        {
            "dataframe.convert_string": True,
            "dataframe.backend": "cudf",
        }
    ):
        ddf = dd.from_dict(
            {
                "x": [1, 2, 3, 4],
                "y": [5.0, 6.0, 7.0, 8.0],
                "z": ["foo", "bar", "ricky", "bobby"],
            },
            npartitions=2,
        )
```

I realize that the string conversion is only meant for the "pandas" backend, but it would be nice for the user not to need to check both these configurations when choosing CPU/GPU. I'm still trying to think of the best way to approach this.

Reply:
Do you have a traceback you can post?

Reply:
<details>
<summary>End of backtrace here</summary>
<br>

```
dask/dataframe/core.py:938: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function to_pyarrow_string at 0x7f0c7ded15a0>
meta = Empty DataFrame
Columns: [x, y, z]
Index: [], enforce_metadata = True
transform_divisions = True, align_dataframes = True
args = [<dask_cudf.DataFrame | 2 tasks | 2 npartitions>], kwargs = {}
name = 'to_pyarrow_string-9709afe6714f5f151cb9064f71cc92a2', parent_meta = None
token = '9709afe6714f5f151cb9064f71cc92a2'
_maybe_align_partitions = <function _maybe_align_partitions at 0x7f0c7dea8670>
dfs = [<dask_cudf.DataFrame | 2 tasks | 2 npartitions>]

    @insert_meta_param_description
    def map_partitions(
        func,
        *args,
        meta=no_default,
        enforce_metadata=True,
        transform_divisions=True,
        align_dataframes=True,
        **kwargs,
    ):
        """Apply Python function on each DataFrame partition.
    
        Parameters
        ----------
        func : function
            Function applied to each partition.
        args, kwargs :
            Arguments and keywords to pass to the function.  At least one of the
            args should be a Dask.dataframe. Arguments and keywords may contain
            ``Scalar``, ``Delayed`` or regular python objects. DataFrame-like args
            (both dask and pandas) will be repartitioned to align (if necessary)
            before applying the function (see ``align_dataframes`` to control).
        enforce_metadata : bool, default True
            Whether to enforce at runtime that the structure of the DataFrame
            produced by ``func`` actually matches the structure of ``meta``.
            This will rename and reorder columns for each partition,
            and will raise an error if this doesn't work,
            but it won't raise if dtypes don't match.
        transform_divisions : bool, default True
            Whether to apply the function onto the divisions and apply those
            transformed divisions to the output.
        align_dataframes : bool, default True
            Whether to repartition DataFrame- or Series-like args
            (both dask and pandas) so their divisions align before applying
            the function. This requires all inputs to have known divisions.
            Single-partition inputs will be split into multiple partitions.
    
            If False, all inputs must have either the same number of partitions
            or a single partition. Single-partition inputs will be broadcast to
            every partition of multi-partition inputs.
        $META
        """
        name = kwargs.pop("token", None)
        parent_meta = kwargs.pop("parent_meta", None)
    
        assert callable(func)
        if name is not None:
            token = tokenize(meta, *args, **kwargs)
        else:
            name = funcname(func)
            token = tokenize(func, meta, *args, **kwargs)
        name = f"{name}-{token}"
    
        from dask.dataframe.multi import _maybe_align_partitions
    
        if align_dataframes:
            args = _maybe_from_pandas(args)
            try:
                args = _maybe_align_partitions(args)
            except ValueError as e:
                raise ValueError(
                    f"{e}. If you don't want the partitions to be aligned, and are "
                    "calling `map_partitions` directly, pass `align_dataframes=False`."
                ) from e
    
        dfs = [df for df in args if isinstance(df, _Frame)]
    
        meta = _get_meta_map_partitions(args, dfs, func, kwargs, meta, parent_meta)
        if all(isinstance(arg, Scalar) for arg in args):
            layer = {
                (name, 0): (
                    apply,
                    func,
                    (tuple, [(arg._name, 0) for arg in args]),
                    kwargs,
                )
            }
            graph = HighLevelGraph.from_collections(name, layer, dependencies=args)
            return Scalar(graph, name, meta)
    
        args2 = []
        dependencies = []
        for arg in args:
            if isinstance(arg, _Frame):
                args2.append(arg)
                dependencies.append(arg)
                continue
            arg = normalize_arg(arg)
            arg2, collections = unpack_collections(arg)
            if collections:
                args2.append(arg2)
                dependencies.extend(collections)
            else:
                args2.append(arg)
    
        kwargs3 = {}
        simple = True
        for k, v in kwargs.items():
            v = normalize_arg(v)
            v, collections = unpack_collections(v)
            dependencies.extend(collections)
            kwargs3[k] = v
            if collections:
                simple = False
    
        divisions = _get_divisions_map_partitions(
            align_dataframes, transform_divisions, dfs, func, args, kwargs
        )
    
        if has_keyword(func, "partition_info"):
            partition_info = {
                (i,): {"number": i, "division": division}
                for i, division in enumerate(divisions[:-1])
            }
    
            args2.insert(0, BlockwiseDepDict(partition_info))
            orig_func = func
    
            def func(partition_info, *args, **kwargs):
                return orig_func(*args, **kwargs, partition_info=partition_info)
    
        if enforce_metadata:
            dsk = partitionwise_graph(
                apply_and_enforce,
                name,
                *args2,
                dependencies=dependencies,
                _func=func,
                _meta=meta,
                **kwargs3,
            )
        else:
            kwargs4 = kwargs if simple else kwargs3
            dsk = partitionwise_graph(
                func, name, *args2, **kwargs4, dependencies=dependencies
            )
    
        graph = HighLevelGraph.from_collections(name, dsk, dependencies=dependencies)
>       return new_dd_object(graph, name, meta, divisions)

dask/dataframe/core.py:6893: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dsk = HighLevelGraph with 2 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x7f0bd8428e50>
 0. from_pandas-c3c722bb68e9a23601561f48bc382927
 1. to_pyarrow_string-9709afe6714f5f151cb9064f71cc92a2

name = 'to_pyarrow_string-9709afe6714f5f151cb9064f71cc92a2'
meta = Empty DataFrame
Columns: [x, y, z]
Index: [], divisions = (0, 2, 3)
parent_meta = None

    def new_dd_object(dsk, name, meta, divisions, parent_meta=None):
        """Generic constructor for dask.dataframe objects.
    
        Decides the appropriate output class based on the type of `meta` provided.
        """
        if has_parallel_type(meta):
>           return get_parallel_type(meta)(dsk, name, meta, divisions)

dask/dataframe/core.py:8051: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <dask_cudf.DataFrame | 4 tasks | 2 npartitions>
dsk = HighLevelGraph with 2 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x7f0bd8428e50>
 0. from_pandas-c3c722bb68e9a23601561f48bc382927
 1. to_pyarrow_string-9709afe6714f5f151cb9064f71cc92a2

name = 'to_pyarrow_string-9709afe6714f5f151cb9064f71cc92a2'
meta = Empty DataFrame
Columns: [x, y, z]
Index: [], divisions = (0, 2, 3)

    def __init__(self, dsk, name, meta, divisions):
>       super().__init__(dsk, name, meta, divisions)

dask/dataframe/core.py:4645: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = [dtype('int64'), dtype('float64'), dtype('O')], index = ('x', 'y', 'z')
dtype = dtype('O'), name = None, copy = False, fastpath = False
dtype_if_empty = <class 'numpy.float64'>

    def create_series_with_explicit_dtype(
        data: Any = None,
        index: ArrayLike | Index | None = None,
        dtype: Dtype | None = None,
        name: str | None = None,
        copy: bool = False,
        fastpath: bool = False,
        dtype_if_empty: Dtype = object,
    ) -> Series:
        """
        Helper to pass an explicit dtype when instantiating an empty Series.
    
        This silences a DeprecationWarning described in GitHub-17261.
    
        Parameters
        ----------
        data : Mirrored from Series.__init__
        index : Mirrored from Series.__init__
        dtype : Mirrored from Series.__init__
        name : Mirrored from Series.__init__
        copy : Mirrored from Series.__init__
        fastpath : Mirrored from Series.__init__
        dtype_if_empty : str, numpy.dtype, or ExtensionDtype
            This dtype will be passed explicitly if an empty Series will
            be instantiated.
    
        Returns
        -------
        Series
        """
        from pandas.core.series import Series
    
        if is_empty_data(data) and dtype is None:
            dtype = dtype_if_empty
>       return Series(
            data=data, index=index, dtype=dtype, name=name, copy=copy, fastpath=fastpath
        )

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/construction.py:915: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'Series' object has no attribute '_mgr'") raised in repr()] Series object at 0x7f0bd81da440>
data = [dtype('int64'), dtype('float64'), dtype('O')], index = ('x', 'y', 'z')
dtype = dtype('O'), name = None, copy = False, fastpath = False

    def __init__(
        self,
        data=None,
        index=None,
        dtype: Dtype | None = None,
        name=None,
        copy: bool = False,
        fastpath: bool = False,
    ) -> None:
    
        if (
            isinstance(data, (SingleBlockManager, SingleArrayManager))
            and index is None
            and dtype is None
            and copy is False
        ):
            # GH#33357 called with just the SingleBlockManager
            NDFrame.__init__(self, data)
            if fastpath:
                # e.g. from _box_col_values, skip validation of name
                object.__setattr__(self, "_name", name)
            else:
                self.name = name
            return
    
        # we are called internally, so short-circuit
        if fastpath:
    
            # data is an ndarray, index is defined
            if not isinstance(data, (SingleBlockManager, SingleArrayManager)):
                manager = get_option("mode.data_manager")
                if manager == "block":
                    data = SingleBlockManager.from_array(data, index)
                elif manager == "array":
                    data = SingleArrayManager.from_array(data, index)
            if copy:
                data = data.copy()
            if index is None:
                index = data.index
    
        else:
    
            name = ibase.maybe_extract_name(name, data, type(self))
    
            if is_empty_data(data) and dtype is None:
                # gh-17261
                warnings.warn(
                    "The default dtype for empty Series will be 'object' instead "
                    "of 'float64' in a future version. Specify a dtype explicitly "
                    "to silence this warning.",
                    FutureWarning,
                    stacklevel=find_stack_level(),
                )
                # uncomment the line below when removing the FutureWarning
                # dtype = np.dtype(object)
    
            if index is not None:
>               index = ensure_index(index)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/series.py:399: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

index_like = ('x', 'y', 'z'), copy = False

    def ensure_index(index_like: Axes, copy: bool = False) -> Index:
        """
        Ensure that we have an index from some index-like object.
    
        Parameters
        ----------
        index_like : sequence
            An Index or other sequence
        copy : bool, default False
    
        Returns
        -------
        index : Index or MultiIndex
    
        See Also
        --------
        ensure_index_from_sequences
    
        Examples
        --------
        >>> ensure_index(['a', 'b'])
        Index(['a', 'b'], dtype='object')
    
        >>> ensure_index([('a', 'a'),  ('b', 'c')])
        Index([('a', 'a'), ('b', 'c')], dtype='object')
    
        >>> ensure_index([['a', 'a'], ['b', 'c']])
        MultiIndex([('a', 'b'),
                ('a', 'c')],
               )
        """
        if isinstance(index_like, Index):
            if copy:
                index_like = index_like.copy()
            return index_like
    
        if isinstance(index_like, ABCSeries):
            name = index_like.name
            return Index._with_infer(index_like, name=name, copy=copy)
    
        if is_iterator(index_like):
            index_like = list(index_like)
    
        if isinstance(index_like, list):
            if type(index_like) is not list:
                # must check for exactly list here because of strict type
                # check in clean_index_list
                index_like = list(index_like)
    
            if len(index_like) and lib.is_all_arraylike(index_like):
                from pandas.core.indexes.multi import MultiIndex
    
                return MultiIndex.from_arrays(index_like)
            else:
                return Index._with_infer(index_like, copy=copy, tupleize_cols=False)
        else:
>           return Index._with_infer(index_like, copy=copy)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/indexes/base.py:7333: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.indexes.base.Index'>, args = (('x', 'y', 'z'),)
kwargs = {'copy': False}

    @classmethod
    def _with_infer(cls, *args, **kwargs):
        """
        Constructor that uses the 1.0.x behavior inferring numeric dtypes
        for ndarray[object] inputs.
        """
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", ".*the Index constructor", FutureWarning)
>           result = cls(*args, **kwargs)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/indexes/base.py:716: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.indexes.base.Index'>, data = ('x', 'y', 'z')
dtype = dtype('O'), copy = False, name = None, tupleize_cols = True, kwargs = {}
PandasArray = <class 'pandas.core.arrays.numpy_.PandasArray'>
RangeIndex = <class 'pandas.core.indexes.range.RangeIndex'>, data_dtype = None

    def __new__(
        cls, data=None, dtype=None, copy=False, name=None, tupleize_cols=True, **kwargs
    ) -> Index:
    
        if kwargs:
            warnings.warn(
                "Passing keywords other than 'data', 'dtype', 'copy', 'name', "
                "'tupleize_cols' is deprecated and will raise TypeError in a "
                "future version.  Use the specific Index subclass directly instead.",
                FutureWarning,
                stacklevel=find_stack_level(),
            )
    
        from pandas.core.arrays import PandasArray
        from pandas.core.indexes.range import RangeIndex
    
        name = maybe_extract_name(name, data, cls)
    
        if dtype is not None:
            dtype = pandas_dtype(dtype)
        if "tz" in kwargs:
            tz = kwargs.pop("tz")
            validate_tz_from_dtype(dtype, tz)
            dtype = tz_to_dtype(tz)
    
        if type(data) is PandasArray:
            # ensure users don't accidentally put a PandasArray in an index,
            #  but don't unpack StringArray
            data = data.to_numpy()
        if isinstance(dtype, PandasDtype):
            dtype = dtype.numpy_dtype
    
        data_dtype = getattr(data, "dtype", None)
    
        # range
        if isinstance(data, (range, RangeIndex)):
            result = RangeIndex(start=data, copy=copy, name=name)
            if dtype is not None:
                return result.astype(dtype, copy=False)
            return result
    
        elif is_ea_or_datetimelike_dtype(dtype):
            # non-EA dtype indexes have special casting logic, so we punt here
            klass = cls._dtype_to_subclass(dtype)
            if klass is not Index:
                return klass(data, dtype=dtype, copy=copy, name=name, **kwargs)
    
            ea_cls = dtype.construct_array_type()
            data = ea_cls._from_sequence(data, dtype=dtype, copy=copy)
            disallow_kwargs(kwargs)
            return Index._simple_new(data, name=name)
    
        elif is_ea_or_datetimelike_dtype(data_dtype):
            data_dtype = cast(DtypeObj, data_dtype)
            klass = cls._dtype_to_subclass(data_dtype)
            if klass is not Index:
                result = klass(data, copy=copy, name=name, **kwargs)
                if dtype is not None:
                    return result.astype(dtype, copy=False)
                return result
            elif dtype is not None:
                # GH#45206
                data = data.astype(dtype, copy=False)
    
            disallow_kwargs(kwargs)
            data = extract_array(data, extract_numpy=True)
            return Index._simple_new(data, name=name)
    
        # index-like
        elif (
            isinstance(data, Index)
            and data._is_backward_compat_public_numeric_index
            and dtype is None
        ):
            return data._constructor(data, name=name, copy=copy)
        elif isinstance(data, (np.ndarray, Index, ABCSeries)):
    
            if isinstance(data, ABCMultiIndex):
                data = data._values
    
            if dtype is not None:
                # we need to avoid having numpy coerce
                # things that look like ints/floats to ints unless
                # they are actually ints, e.g. '0' and 0.0
                # should not be coerced
                # GH 11836
                data = sanitize_array(data, None, dtype=dtype, copy=copy)
    
                dtype = data.dtype
    
            if data.dtype.kind in ["i", "u", "f"]:
                # maybe coerce to a sub-class
                arr = data
            elif data.dtype.kind in ["b", "c"]:
                # No special subclass, and Index._ensure_array won't do this
                #  for us.
                arr = np.asarray(data)
            else:
                arr = com.asarray_tuplesafe(data, dtype=_dtype_obj)
    
                if dtype is None:
                    arr = _maybe_cast_data_without_dtype(
                        arr, cast_numeric_deprecated=True
                    )
                    dtype = arr.dtype
    
                    if kwargs:
                        return cls(arr, dtype, copy=copy, name=name, **kwargs)
    
            klass = cls._dtype_to_subclass(arr.dtype)
            arr = klass._ensure_array(arr, dtype, copy)
            disallow_kwargs(kwargs)
            return klass._simple_new(arr, name)
    
        elif is_scalar(data):
            raise cls._scalar_data_error(data)
        elif hasattr(data, "__array__"):
            return Index(np.asarray(data), dtype=dtype, copy=copy, name=name, **kwargs)
        else:
    
            if tupleize_cols and is_list_like(data):
                # GH21470: convert iterable to list before determining if empty
                if is_iterator(data):
                    data = list(data)
    
                if data and all(isinstance(e, tuple) for e in data):
                    # we must be all tuples, otherwise don't construct
                    # 10697
                    from pandas.core.indexes.multi import MultiIndex
    
                    return MultiIndex.from_tuples(
                        data, names=name or kwargs.get("names")
                    )
            # other iterable of some kind
    
            subarr = com.asarray_tuplesafe(data, dtype=_dtype_obj)
            if dtype is None:
                # with e.g. a list [1, 2, 3] casting to numeric is _not_ deprecated
                subarr = _maybe_cast_data_without_dtype(
                    subarr, cast_numeric_deprecated=False
                )
                dtype = subarr.dtype
>           return Index(subarr, dtype=dtype, copy=copy, name=name, **kwargs)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/indexes/base.py:572: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.indexes.base.Index'>
data = array(['x', 'y', 'z'], dtype=object), dtype = dtype('O'), copy = False
name = None, tupleize_cols = True, kwargs = {}
PandasArray = <class 'pandas.core.arrays.numpy_.PandasArray'>
RangeIndex = <class 'pandas.core.indexes.range.RangeIndex'>
data_dtype = dtype('O')

    def __new__(
        cls, data=None, dtype=None, copy=False, name=None, tupleize_cols=True, **kwargs
    ) -> Index:
    
        if kwargs:
            warnings.warn(
                "Passing keywords other than 'data', 'dtype', 'copy', 'name', "
                "'tupleize_cols' is deprecated and will raise TypeError in a "
                "future version.  Use the specific Index subclass directly instead.",
                FutureWarning,
                stacklevel=find_stack_level(),
            )
    
        from pandas.core.arrays import PandasArray
        from pandas.core.indexes.range import RangeIndex
    
        name = maybe_extract_name(name, data, cls)
    
        if dtype is not None:
            dtype = pandas_dtype(dtype)
        if "tz" in kwargs:
            tz = kwargs.pop("tz")
            validate_tz_from_dtype(dtype, tz)
            dtype = tz_to_dtype(tz)
    
        if type(data) is PandasArray:
            # ensure users don't accidentally put a PandasArray in an index,
            #  but don't unpack StringArray
            data = data.to_numpy()
        if isinstance(dtype, PandasDtype):
            dtype = dtype.numpy_dtype
    
        data_dtype = getattr(data, "dtype", None)
    
        # range
        if isinstance(data, (range, RangeIndex)):
            result = RangeIndex(start=data, copy=copy, name=name)
            if dtype is not None:
                return result.astype(dtype, copy=False)
            return result
    
        elif is_ea_or_datetimelike_dtype(dtype):
            # non-EA dtype indexes have special casting logic, so we punt here
            klass = cls._dtype_to_subclass(dtype)
            if klass is not Index:
                return klass(data, dtype=dtype, copy=copy, name=name, **kwargs)
    
            ea_cls = dtype.construct_array_type()
            data = ea_cls._from_sequence(data, dtype=dtype, copy=copy)
            disallow_kwargs(kwargs)
            return Index._simple_new(data, name=name)
    
        elif is_ea_or_datetimelike_dtype(data_dtype):
            data_dtype = cast(DtypeObj, data_dtype)
            klass = cls._dtype_to_subclass(data_dtype)
            if klass is not Index:
                result = klass(data, copy=copy, name=name, **kwargs)
                if dtype is not None:
                    return result.astype(dtype, copy=False)
                return result
            elif dtype is not None:
                # GH#45206
                data = data.astype(dtype, copy=False)
    
            disallow_kwargs(kwargs)
            data = extract_array(data, extract_numpy=True)
            return Index._simple_new(data, name=name)
    
        # index-like
        elif (
            isinstance(data, Index)
            and data._is_backward_compat_public_numeric_index
            and dtype is None
        ):
            return data._constructor(data, name=name, copy=copy)
        elif isinstance(data, (np.ndarray, Index, ABCSeries)):
    
            if isinstance(data, ABCMultiIndex):
                data = data._values
    
            if dtype is not None:
                # we need to avoid having numpy coerce
                # things that look like ints/floats to ints unless
                # they are actually ints, e.g. '0' and 0.0
                # should not be coerced
                # GH 11836
>               data = sanitize_array(data, None, dtype=dtype, copy=copy)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/indexes/base.py:516: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = array(['x', 'y', 'z'], dtype=object), index = None, dtype = dtype('O')
copy = False, raise_cast_failure = True

    def sanitize_array(
        data,
        index: Index | None,
        dtype: DtypeObj | None = None,
        copy: bool = False,
        raise_cast_failure: bool = True,
        *,
        allow_2d: bool = False,
    ) -> ArrayLike:
        """
        Sanitize input data to an ndarray or ExtensionArray, copy if specified,
        coerce to the dtype if specified.
    
        Parameters
        ----------
        data : Any
        index : Index or None, default None
        dtype : np.dtype, ExtensionDtype, or None, default None
        copy : bool, default False
        raise_cast_failure : bool, default True
        allow_2d : bool, default False
            If False, raise if we have a 2D Arraylike.
    
        Returns
        -------
        np.ndarray or ExtensionArray
    
        Notes
        -----
        raise_cast_failure=False is only intended to be True when called from the
        DataFrame constructor, as the dtype keyword there may be interpreted as only
        applying to a subset of columns, see GH#24435.
        """
        if isinstance(data, ma.MaskedArray):
            data = sanitize_masked_array(data)
    
        if isinstance(dtype, PandasDtype):
            # Avoid ending up with a PandasArray
            dtype = dtype.numpy_dtype
    
        # extract ndarray or ExtensionArray, ensure we have no PandasArray
>       data = extract_array(data, extract_numpy=True, extract_range=True)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/construction.py:534: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = array(['x', 'y', 'z'], dtype=object), extract_numpy = True
extract_range = True

    def extract_array(
        obj: T, extract_numpy: bool = False, extract_range: bool = False
    ) -> T | ArrayLike:
        """
        Extract the ndarray or ExtensionArray from a Series or Index.
    
        For all other types, `obj` is just returned as is.
    
        Parameters
        ----------
        obj : object
            For Series / Index, the underlying ExtensionArray is unboxed.
    
        extract_numpy : bool, default False
            Whether to extract the ndarray from a PandasArray.
    
        extract_range : bool, default False
            If we have a RangeIndex, return range._values if True
            (which is a materialized integer ndarray), otherwise return unchanged.
    
        Returns
        -------
        arr : object
    
        Examples
        --------
        >>> extract_array(pd.Series(['a', 'b', 'c'], dtype='category'))
        ['a', 'b', 'c']
        Categories (3, object): ['a', 'b', 'c']
    
        Other objects like lists, arrays, and DataFrames are just passed through.
    
        >>> extract_array([1, 2, 3])
        [1, 2, 3]
    
        For an ndarray-backed Series / Index the ndarray is returned.
    
        >>> extract_array(pd.Series([1, 2, 3]))
        array([1, 2, 3])
    
        To extract all the way down to the ndarray, pass ``extract_numpy=True``.
    
        >>> extract_array(pd.Series([1, 2, 3]), extract_numpy=True)
        array([1, 2, 3])
        """
>       if isinstance(obj, (ABCIndex, ABCSeries)):

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/construction.py:444: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.dtypes.generic.ABCBase'>
inst = array(['x', 'y', 'z'], dtype=object)

    @classmethod  # type: ignore[misc]
    def _instancecheck(cls, inst) -> bool:
>       return _check(inst) and not isinstance(inst, type)

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/dtypes/generic.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

inst = array(['x', 'y', 'z'], dtype=object)

    def _check(inst):
>       return getattr(inst, attr, "_typ") in comp
E       RecursionError: maximum recursion depth exceeded while calling a Python object

/datasets/rzamora/miniconda3/envs/cudf_2304_dev/lib/python3.10/site-packages/pandas/core/dtypes/generic.py:41: RecursionError
!!! Recursion error detected, but an error occurred locating the origin of recursion.
  The following exception happened when comparing locals in the stack frame:
    ValueError: The truth value of a DataFrame is ambiguous. Use a.any() or a.all().
  Displaying first and last 10 stack frames out of 796.

The above exception was the direct cause of the following exception:

args = ({'x': [1, 2, 3, 4], 'y': [5.0, 6.0, 7.0, 8.0], 'z': ['foo', 'bar', 'ricky', 'bobby']},)
kwargs = {'npartitions': 2}
func = <function CudfBackendEntrypoint.from_dict at 0x7f0bd83f2b00>

    @wraps(fn)
    def wrapper(*args, **kwargs):
        func = getattr(self, dispatch_name)
        try:
>           return func(*args, **kwargs)

dask/backends.py:133: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = {'x': [1, 2, 3, 4], 'y': [5.0, 6.0, 7.0, 8.0], 'z': ['foo', 'bar', 'ricky', 'bobby']}
npartitions = 2, orient = 'columns', dtype = None, columns = None
constructor = <class 'cudf.core.dataframe.DataFrame'>

    @staticmethod
    def from_dict(
        data,
        npartitions,
        orient="columns",
        dtype=None,
        columns=None,
        constructor=cudf.DataFrame,
    ):
    
>       return _default_backend(
            dd.from_dict,
            data,
            npartitions=npartitions,
            orient=orient,
            dtype=dtype,
            columns=columns,
            constructor=constructor,
        )

../cudf/python/dask_cudf/dask_cudf/backends.py:526: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function from_dict at 0x7f0c7de50790>
args = ({'x': [1, 2, 3, 4], 'y': [5.0, 6.0, 7.0, 8.0], 'z': ['foo', 'bar', 'ricky', 'bobby']},)
kwargs = {'columns': None, 'constructor': <class 'cudf.core.dataframe.DataFrame'>, 'dtype': None, 'npartitions': 2, ...}

    def _default_backend(func, *args, **kwargs):
        # Utility to call a dask.dataframe function with
        # the default ("pandas") backend
    
        # NOTE: Some `CudfBackendEntrypoint` methods need to
        # invoke the "pandas"-version of the same method, but
        # with custom kwargs (e.g. `engine`). In these cases,
        # an explicit "pandas" config context is needed to
        # avoid a recursive loop
        with config.set({"dataframe.backend": "pandas"}):
>           return func(*args, **kwargs)

../cudf/python/dask_cudf/dask_cudf/backends.py:450: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ({'x': [1, 2, 3, 4], 'y': [5.0, 6.0, 7.0, 8.0], 'z': ['foo', 'bar', 'ricky', 'bobby']},)
kwargs = {'columns': None, 'constructor': <class 'cudf.core.dataframe.DataFrame'>, 'dtype': None, 'npartitions': 2, ...}
func = <function from_dict at 0x7f0c7de50700>

    @wraps(fn)
    def wrapper(*args, **kwargs):
        func = getattr(self, dispatch_name)
        try:
            return func(*args, **kwargs)
        except Exception as e:
>           raise type(e)(
                f"An error occurred while calling the {funcname(func)} "
                f"method registered to the {self.backend} backend.\n"
                f"Original Message: {e}"
            ) from e
E           RecursionError: An error occurred while calling the from_dict method registered to the pandas backend.
E           Original Message: maximum recursion depth exceeded while calling a Python object

dask/backends.py:135: RecursionError

The above exception was the direct cause of the following exception:

backend = 'cudf', convert = True

    @pytest.mark.skipif(not PYARROW_STRINGS_AVAILABLE, reason="Requires pyarrow strings")
    @pytest.mark.parametrize("backend", ["cudf", "pandas"])
    @pytest.mark.parametrize("convert", [True])
    def test_from_pandas_convert_string_config_cudf(backend, convert):
    
        with dask.config.set(
            {
                "dataframe.convert_string": convert,
                "dataframe.backend": backend,
            }
        ):
>           ddf = dd.from_dict(
                {
                    "x": [1, 2, 3, 4],
                    "y": [5.0, 6.0, 7.0, 8.0],
                    "z": ["foo", "bar", "ricky", "bobby"],
                },
                npartitions=2,
            )

dask/dataframe/io/tests/test_io.py:290: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ({'x': [1, 2, 3, 4], 'y': [5.0, 6.0, 7.0, 8.0], 'z': ['foo', 'bar', 'ricky', 'bobby']},)
kwargs = {'npartitions': 2}
func = <function CudfBackendEntrypoint.from_dict at 0x7f0bd83f2b00>

    @wraps(fn)
    def wrapper(*args, **kwargs):
        func = getattr(self, dispatch_name)
        try:
            return func(*args, **kwargs)
        except Exception as e:
>           raise type(e)(
                f"An error occurred while calling the {funcname(func)} "
                f"method registered to the {self.backend} backend.\n"
                f"Original Message: {e}"
            ) from e
E           RecursionError: An error occurred while calling the from_dict method registered to the cudf backend.
E           Original Message: An error occurred while calling the from_dict method registered to the pandas backend.
E           Original Message: maximum recursion depth exceeded while calling a Python object

dask/backends.py:135: RecursionError
```

</details>

The cause of the failure is just that a cudf DataFrame/Series/Index will be recognized as dataframelike/serieslike/indexlike, and the cudf string dtype will be recognized as an object string type. Therefore, you will end up trying to convert the data to pyarrow strings, but that will fail when you try to pass gpu buffers into pandas.

Reply:
Brief explanation of the error in [another thread](https://github.com/dask/dask/pull/9926#discussion_r1108914888)